### DISCUSSION TITLE
How to solve 'Error: require() of ES Module xxx not supported. Instead change the require of xxx to a dynamic import() which is available in all CommonJS modules.'?

### DISCUSSION THREAD

**Discussion by foxaltus** [rollup/rollup#5629] [Q&A]
Hi there!

I'm adding Vitest (v2) to one of my legacy react app (webpack + commonjs) and I receive this error when running the tests:

```
Error: require() of ES Module xxx\my_library\node_modules\.pnpm\d3@7.9.0\node_modules\d3\src\index.js from xxx\my_library\lib\index.js not supported.
Instead change the require of xxx\my_library\node_modules\.pnpm\d3@7.9.0\node_modules\d3\src\index.js in xxx\my_library\lib\index.js to a dynamic import() which is available in all CommonJS modules.
 ❯ ../my_library/lib/index.js:11:10
      9| var useMediaQuery = require('@mui/material/useMediaQuery');
     10| var React = require('react');
     11| var d3 = require('d3');
       |          ^
```

`my_library` is my own package, and it's built with rollup + commonjs plugin. What option can I use to make it output dynamic imports instead of requires when importing an ES Module?

**Comment by foxaltus** [comment]
For those reading this, I was not able to fix this issue with `rollup`. I switched to `tsup` and it solved the problem (kudos to Matt Pocock for this guide: https://www.totaltypescript.com/how-to-create-an-npm-package).

My final `tsconfig.json`:
```
{
  "compilerOptions": {
    "esModuleInterop": true,
    "skipLibCheck": true,
    "target": "es2022",
    "allowJs": true,
    "resolveJsonModule": true,
    "moduleDetection": "force",
    "isolatedModules": true,
    "strict": true,
    "noImplicitOverride": true,
    "module": "Preserve", // because we're using tsup to build
    "noEmit": true,
    "jsx": "react-jsx",
    "types": ["vite/client"]
  },
  "include": ["src"]
}
```

with `tsup.config.ts`:
```
import { defineConfig } from 'tsup';

export default defineConfig({
  entryPoints: ['src/index.tsx'],
  format: ['cjs', 'esm'],
  dts: true,
  outDir: 'dist',
  clean: true,
  loader: {
    '.png': 'dataurl',
    '.woff': 'dataurl',
    '.woff2': 'dataurl',
  },
});
```

and `package.json`:

```
{
  ...
  "type": "module",
  "main": "dist/index.js",
  "files": [
    "dist"
  ],
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": "./dist/index.js",
      "default": "./dist/index.cjs"
    }
  },
  "scripts": {
    "build": "tsup",
    ... 
  },
  "devDependencies": {
    ...
    "tsup": "^8.2.4",
    "typescript": "^5.5.4",
  }
  ...
}
```

**Comment by chrisjcthomas** [comment]
This issue occurs because `d3` (since version 7) is a pure ES Module (ESM) package. A CommonJS bundle (like your library's CJS build output) cannot use standard `require()` to load a pure ESM module synchronously in Node.js.

Since your library bundles its files, you can resolve this using one of the following approaches:

### 1. Bundle ESM Dependencies into your CJS output (Recommended)
Instead of marking `d3` as an external dependency in your Rollup configuration, let Rollup bundle it directly. Since Rollup compiles ESM, it will resolve and compile `d3`'s exports directly into your CJS output.

In your `rollup.config.js`, make sure `d3` is **not** included in the `external` array, and configure `@rollup/plugin-node-resolve` to resolve node modules:

```javascript
import resolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';

export default {
  input: 'src/index.js',
  output: {
    file: 'lib/index.js',
    format: 'cjs',
    exports: 'auto'
  },
  plugins: [
    resolve(), // Resolves third-party modules in node_modules
    commonjs() // Converts CommonJS modules to ES6
  ],
  // Keep peer dependencies external, but let d3 be bundled
  external: ['react', 'react-dom', '@mui/material']
};
```

### 2. Output both ESM and CommonJS
Modern libraries should distribute ES Modules alongside CommonJS. This allows modern tools (like Vitest) to import the ESM version, bypassing Node's CJS restriction:

```javascript
output: [
  { file: 'lib/index.js', format: 'cjs' },
  { file: 'lib/index.mjs', format: 'es' }
]
```
Ensure your `package.json` lists both entry points:
```json
{
  "main": "lib/index.js",
  "module": "lib/index.mjs"
}
```
